frontend/catalyst/jax_extras/tracing.py

```diff
@@ -248,6 +312,97 @@ def __init__(self, boxid: int, e: JaxprEqn):
     return [b.e for b in stable_toposort(boxes)]  # [4]
 
 
+@transformation_with_aux
+def expanded_fun(static_args, *args_expanded):
+    """Function transformation making the function to accept its arguments in the expanded format
+    (with the dimension variables added)"""
+    (in_type, expansion_strategy) = static_args
+    args_collapsed = [a for a, (_, k) in zip(args_expanded, in_type) if k]
+    res_flat = yield args_collapsed, {}
+    num_implicit_inputs = len([() for _, k in in_type if not k])
+    all_outs, out_sig = infer_output_type_python(
+        args_expanded, res_flat, expansion_strategy, num_implicit_inputs
+    )
+    yield all_outs, out_sig
+
+
+@dataclass
+class InputSignature:
+    """Meta-parameters of a function which are available before the tracing to the function
+    starts."""
+
+    in_type: InputType
+    in_tree: PyTreeDef
+    in_expanded_args: List[DynamicJaxprTracer]
+
+    def num_implicit_inputs(self) -> int:
+        """Return the number of implicit input arguments of the function"""
+        return len([() for _, k in self.in_type if not k])
```

Comment:
Minor comment: I prefer to use filter and reduce here. And if possible we can cache the result. I know it is made as a dataclass because the state won't change. (Or at least that's what I believe). I would prefer something like:

```python
def __init__(in_type, in_tree, in_expanded_args):
   self.num_implicit = reduce(filter(...))
```

Reply:
Do you mean something like this?
``` python
reduce(lambda acc, _: acc+1, filter(lambda x: x[1] is explicit, typ), 0)
```
I found standard Python functools undeveloped, so I would prefer continue using list comprehensions, even with a slight performance decrease. What I would agree to see in the code base would look like
`reduce( add(const(1)), filter((==explicit).fst), typ, 0)` but afaik Python stdlib lacks these simple functions and currying is also not working this way.